core/src/telemetry/metrics.rs

```diff
@@ -637,44 +628,66 @@ pub fn start_prometheus_metric_exporter(
 
 /// Buffers [MetricEvent]s for periodic consumption by lang
 #[derive(Debug)]
-pub struct MetricsCallBuffer {
-    instrument_ids: AtomicU64,
-    attribute_ids: AtomicU64,
-    calls_rx: crossbeam::channel::Receiver<MetricEvent>,
-    calls_tx: crossbeam::channel::Sender<MetricEvent>,
+pub struct MetricsCallBuffer<I> {
+    calls_rx: crossbeam::channel::Receiver<MetricEvent<I>>,
+    calls_tx: crossbeam::channel::Sender<MetricEvent<I>>,
 }
-impl MetricsCallBuffer {
+
+impl<I> MetricsCallBuffer<I>
+where
+    I: Clone,
+{
     /// Create a new buffer with the given capacity
     pub fn new(buffer_size: usize) -> Self {
         let (calls_tx, calls_rx) = crossbeam::channel::bounded(buffer_size);
-        MetricsCallBuffer {
-            instrument_ids: AtomicU64::new(0),
-            attribute_ids: AtomicU64::new(0),
-            calls_rx,
-            calls_tx,
-        }
+        MetricsCallBuffer { calls_rx, calls_tx }
     }
-    fn new_instrument(&self, params: MetricParameters, kind: MetricKind) -> BufferInstrument {
-        let id = self.instrument_ids.fetch_add(1, Ordering::AcqRel);
-        let _ = self.calls_tx.send(MetricEvent::Create { params, id, kind });
+    fn new_instrument(&self, params: MetricParameters, kind: MetricKind) -> BufferInstrument<I> {
+        let hole = LazyBufferInstrument::hole();
+        let _ = self.calls_tx.send(MetricEvent::Create {
+            params,
+            kind,
+            populate_into: hole.clone(),
+        });
         BufferInstrument {
             kind,
-            id,
+            instrument_ref: hole,
             tx: self.calls_tx.clone(),
         }
     }
 }
-impl CoreMeter for MetricsCallBuffer {
-    fn new_attributes(&self, opts: MetricsAttributesOptions) -> MetricAttributes {
-        let id = self.attribute_ids.fetch_add(1, Ordering::AcqRel);
+
+impl<I> CoreMeter for MetricsCallBuffer<I>
+where
+    I: Debug + Send + Sync + Clone + 'static,
```

Comment:
Hrmm, so I guess if I'm tying my non-cloneable Python metric's instrument instance to this, I will have to wrap it in Arc or similar? It doesn't seem like `CustomMetricAttributes` has to implement `Clone`. I don't think it's a problem though.

Reply:
The trait doesn't require it, but the places where it's used do (when they need it) which is fairly typical practice